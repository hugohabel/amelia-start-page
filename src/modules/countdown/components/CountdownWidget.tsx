// Internal Dependencies
import { CountdownFormat } from '../../../common/types/amelia';
import {
  getDiffUntilDate,
  formatTimeDiff,
  isDateInThePast,
} from '../utils/timeCalculations/timeCalculations';
import { log } from '../../../common/utils/logs/logs';

// Styles
import styles from './CountdownWidget.module.css';
import sharedStyles from '../../../common/styles/Shared.module.css';

// Interfaces + Types
interface ICountdownWidgetProps {
  eventName: string;
  eventDate: string;
  eventFormat: CountdownFormat;
  eventEmoji: string;
}

/**
 * Countdown Widget
 *
 * @param params ICountdownWidgetProps
 * @returns JSX
 */
function CountdownWidget({ eventDate, eventName, eventFormat, eventEmoji }: ICountdownWidgetProps) {
  let isEventInThePast = false;
  let formattedOutput = '';

  try {
    const timeDiff = getDiffUntilDate(eventDate);
    isEventInThePast = isDateInThePast(timeDiff);
    formattedOutput = formatTimeDiff(timeDiff, eventFormat);
  } catch (err: any) {
    log({
      level: 'error',
      message: err.message,
      data: `eventDate: ${eventDate} - eventFormat: ${eventFormat}`,
      file: 'CountdownWidget.ts',
      location: 'CountdownWidget',
    });
  }

  return (
    <div className={styles.countdownWidget}>
      <h3>{`${eventEmoji} ${eventName}`}</h3>
      <p className={styles.separatorWord}>{!isEventInThePast && `in`}</p>
      <h2 className={sharedStyles.mainText}>{formattedOutput}</h2>
    </div>
  );
}

export { CountdownWidget };
