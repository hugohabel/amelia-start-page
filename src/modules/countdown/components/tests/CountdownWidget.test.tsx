// External Dependencies
import React from 'react';
import { render } from '@testing-library/react';

// Internal Dependencies
import { CountdownWidget } from '../CountdownWidget';

describe('Countdown Widget', () => {
  // Setup and Teardown

  beforeAll(() => {
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date(2021, 5, 13));
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('matches snapshot', async () => {
    const component = render(
      <CountdownWidget
        eventName="Christmas"
        eventDate="2021-12-25T05:00:00.000Z"
        eventFormat="days+hours"
        eventEmoji="ðŸ’¡"
      />
    );

    expect(component.asFragment()).toMatchSnapshot();
  });
});
