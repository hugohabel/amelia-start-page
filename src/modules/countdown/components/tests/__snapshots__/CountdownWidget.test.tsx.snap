// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown Widget matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="countdownWidget"
  >
    <h3>
      Christmas
    </h3>
    <p
      class="separatorWord"
    >
      in
    </p>
    <h2
      class="remainingTime"
    >
      195 days 0 hours
    </h2>
  </div>
</DocumentFragment>
`;
